website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md

```diff
@@ -0,0 +1,208 @@
+    sdktrace "go.opentelemetry.io/otel/sdk/trace"
+    semconv "go.opentelemetry.io/otel/semconv/v1.21.0"
+)
+
+func main() {
+    exporter := otlptracehttp.NewUnstarted(
+        otlptracehttp.WithEndpointURL("http://127.0.0.1:4318/v1/traces"),
+        otlptracehttp.WithTimeout(10*time.Second),
+    )
+
+    tp := sdktrace.NewTracerProvider(
+        sdktrace.WithBatcher(exporter),
+        sdktrace.WithResource(sdkresource.NewSchemaless(
+            semconv.ServiceName("test-app"),
+        )),
+    )
+
+    otel.SetTracerProvider(tp)
+
+    defer func() {
+        shutdownCtx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+        defer cancel()
+
+        _ = tp.Shutdown(shutdownCtx)
+    }()
+
+    tracer := otel.Tracer("")
+
+    ctx, span := tracer.Start(context.Background(), "main")
+    defer span.End()
+
+    clientOpts := options.Client().ApplyURI("mongodb://localhost:27017")
+    client, err := mongo.Connect(ctx, clientOpts)
+
+    if err != nil {
+        log.Panicf("Failed to connect to FerretDB: %v", err)
+    }
+
+    defer func() {
+        _ = client.Disconnect(ctx)
+    }()
+
+    collection := client.Database("testdb").Collection("customers")
+
+    user := bson.D{
+        {Key: "name", Value: "John Doe"},
+        {Key: "email", Value: "john.doe@example.com"},
+    }
+
+    insertCtx, insertSpan := tracer.Start(ctx, "InsertCustomer")
+    _, _ = collection.InsertOne(insertCtx, user)
+    insertSpan.End()
+
+    findCtx, findSpan := tracer.Start(ctx, "FindCustomer")
+    filter := bson.D{{Key: "name", Value: "John Doe"}}
+
+    var result bson.D
+    _ = collection.FindOne(findCtx, filter).Decode(&result)
+    findSpan.End()
+
+    fmt.Printf("Found document: %v\n", result)
+}
+```
+
+In this setup, we initialize the OpenTelemetry tracer provider and configure it to export traces to a local Jaeger endpoint.
+
+If we run this application, we will see that the spans for the `InsertCustomer` and `FindCustomer` operations are created:
+
+[![Trace without propagation](/img/blog/ferretdb-otel/without-propagation.png)](/img/blog/ferretdb-otel/without-propagation.png)
+
+Let's add more details.
+If we pass the tracing context through the `comment` field, we will see that the spans are linked.
+Let's modify the `FindCustomer` part:
+
+```go
+    findCtx, findSpan := tracer.Start(ctx, "FindCustomer")
+    filter := bson.D{{Key: "name", Value: "John Doe"}}
+
+    traceID := findSpan.SpanContext().TraceID().String()
+    spanID := findSpan.SpanContext().SpanID().String()
+
+    traceContext := struct {
+        TraceID string `json:"traceID"`
+        SpanID  string `json:"spanID"`
+    }{
+        TraceID: traceID,
+        SpanID:  spanID,
+    }
+
+    comment, _ := json.Marshal(map[string]interface{}{
+        "ferretDB": traceContext,
+    })
+
+    var result bson.D
+    _ = collection.FindOne(findCtx, filter, options.FindOne().SetComment(string(comment))).Decode(&result)
+    findSpan.End()
+```
```

Comment:
That code also should be reduced to relevant snippets and made more tights. Right now, code dominates the message of the blog post; it should not.

Reply:
fixed